File: 7th_100/problem603.py
from sage.all import *

N = 10**6
K = 10**12
MOD = 10**9 + 7
def brute_force(n):
    n_str = str(n)
    n_digits = len(n_str)
    ans = 0
    for length in range(1, n_digits+1):
        for start in range(n_digits - length + 1):
            d = int(n_str[start:start+length], 10)
            ans += d
    return ans

#def S(n):
#    ans = 0
#
#    n_str = str(n)
#    n_digits = len(n_str)
#
#    coeff = [0] * n_digits
#    coeff[0] = 1
#    for i in range(1, n_digits):
#        coeff[i] = (pow(10, i, MOD) + coeff[i-1]) % MOD
#
#    for inversed_place_value, digit_str in enumerate(n_str):
#        digit = int(digit_str, 10)
#        place_value = n_digits - inversed_place_value - 1
#        n_appearances_per_place_value = inversed_place_value + 1
#        print(digit, coeff[place_value], n_appearances_per_place_value)
#        ans += digit * coeff[place_value] * n_appearances_per_place_value
#        ans %= MOD
#
#    return ans

def rho(k):
    return (10**(k) - 1)//9

def S(n, n_periods):
    R = IntegerModRing(MOD)
    ans = R(0)

    n_str = str(n)
    period_length = len(n_str)
    n_periods = n_periods
    n_digits = n_periods * period_length

    for inversed_place_value, digit_str in enumerate(n_str):
        if inversed_place_value % 10000 == 0:
            print(inversed_place_value)
        digit = int(digit_str, 10)
        offset = inversed_place_value + 1
        coeff_sum = 0
        #for period_idx in range(n_periods):
            #coeff = 1
            #coeff *= rho(n_digits + 1 - offset - period_idx * period_length)
            #coeff *= offset + period_length * period_idx
            #coeff_sum += coeff
            #coeff = (10**(n_digits + 1 - offset - period_idx * period_length)-1)*(offset+period_idx * period_length)
            #coeff_sum += coeff
        #print("expected coeff", coeff_sum)
        #ans += digit * coeff_sum // 9
        # Since multiplicative_order(10, 10**9+7) = 10**9+6, then L, N, a in 10**(L+1-a) (mod 10**9+7) must be of the ModRing(10**9+6)! Putting L, N, a to IntegerModRing(10**9+7) would result in an incorrect result.
        L = period_length
        assert(isinstance(L, int))
        N = n_periods
        assert(isinstance(N, int))
        a = offset
        assert(isinstance(a, int))
        part_1 = a*((R(10)**(L+1-a))*(R(10)**(L*N)-1) // (R(10)**L-1) - N)
        part_2 = L * (2*(R(10)**(L+1-a))*(R(10)**(L*N)-1) + (1-R(10)**L)*(N*(2*(R(10)**(L+1-a))-R(10)**L+1)) - (N**2)*((R(10)**L-1)**2)) // (2*((R(10)**L-1)**2))
        coeff = part_1 + part_2
        ans += digit * coeff // 9
    return ans

def C(n):
    p = primes_first_n(n)
    p = list(map(str, p))
    ans = "".join(p)
    return ans

if __name__ == "__main__":
    print(S(C(N), K))

File: 7th_100/problem603.cpp
#include<pari/pari.h>
#include<iostream>
#include<string>
#include<algorithm>
#include<vector>
#include<cassert>
#include<sstream>
#include<cmath>

using namespace std;

typedef __int128_t ui;
#define endl "\n"

constexpr ui pow(ui base, ui exp)
{
    if (exp == 0)
        return 1;
    return base * exp;
};

const ui N = pow(10, 6);
const ui K = pow(10, 12);
const ui MOD = pow(10, 9) + 7;

ui pow_mod(const ui& base, const ui& exp, const ui& mod)
{
    ui b = base;
    ui e = exp;
    ui ans = 1;
    while (e > 0)
    {
        if (e % 2 == 1)
        {
            ans *= b;
            ans %= mod;
        }
        e /= 2;
        b *= b;
        b %= mod;
    }
    return ans;
}

string i128_to_string(__int128_t n)
{
    stringstream stream;
    if (n < 0)
        n+=MOD;
    bool sign = false;
    if (n < 0)
    {
        sign = true;
        n = -n;
    }
    while (n > 0)
    {
        stream << char('0' + n % 10);
        n = n / 10;
    }
    if (sign)
        stream << "-";

    string str = stream.str();
    reverse(str.begin(), str.end());

    if (str == "")
        return "0";
    return str;
}

ostream& operator<<(ostream& os, const __int128_t& n)
{
    os << i128_to_string(n);
    return os;
}

vector<ui> get_primes(ui n)
{
    vector<ui> primes;
    primes.reserve(n/10);

    vector<bool> is_prime = vector<bool>(n/2+1, true);
    ui sqrt_n = ui(sqrt(n));
    for (ui p = 3; p <= sqrt_n; p+=2)
    {
        if (!is_prime[(p-3)/2])
            continue;
        for (ui np = 3*p; np <= n; np += 2*p)
            is_prime[(np-3)/2] = false;
    }
    primes.push_back(2);
    ui size = is_prime.size();
    for (ui p = 0; 2*p+3 <= n; p++)
    {
        if (is_prime[p])
            primes.push_back(2*p+3);
    }
    return primes;
}

ui inv_mod(const ui n, const ui mod)
{
    uint64_t t1;
    return Fl_invgen(n, mod, &t1);
}

string concat_nums(const vector<ui>& nums)
{
    stringstream s;
    for (auto n: nums)
        s << n;
    return s.str();
}

ui MVC(const ui& a, const ui& b, const ui& mod) // multiply very carefully
{
    return (a * b) % mod;
}

ui S(const string& pattern, const ui& n_periods)
{
    ui ans = 0;
    ui period_length = pattern.size();
    ui n_digits = MVC(n_periods, period_length, MOD);

    ui L = period_length;
    ui N = n_periods;

    ui pow_10_L = pow_mod(10, L, MOD);
    ui pow_10_LN = pow_mod(10, L*N, MOD);
    ui pow_10_L1_2 = pow_mod((pow_10_L-1) % MOD, 2, MOD);
    ui pow_N_2 = pow_mod(N, 2, MOD);

    ui inv_mod_9 = inv_mod(9, MOD);

    for (ui inversed_place_value = 0; inversed_place_value < period_length; inversed_place_value++)
    {
        if (inversed_place_value % 100000 == 0)
            cout << inversed_place_value << endl;

        ui digit = pattern[inversed_place_value] - '0';
        ui offset = inversed_place_value + 1;
        ui a = offset;

        ui pow_10_La1 = pow_mod(10, L+1-a, MOD);

        ui part_1_a = MVC(pow_10_La1, (pow_10_LN-1)%MOD, MOD);
        ui part_1_b = (pow_10_L-1)%MOD;
        ui inv_mod_part_1_b = inv_mod(part_1_b, MOD);
        ui part_1 = (MVC(part_1_a, inv_mod_part_1_b, MOD) - N) % MOD;

        ui part_2_a = MVC(N, pow_10_L-1, MOD);
        part_2_a = MVC(part_2_a, pow_10_L - MVC(2,pow_10_La1,MOD) - 1, MOD);
        ui part_2_b = MVC(2, MVC(pow_10_La1, (pow_10_LN - 1) % MOD, MOD), MOD);
        ui part_2_c = MVC(pow_N_2, pow_10_L1_2, MOD);
        ui part_2_d = MVC(2, pow_10_L1_2, MOD);
        ui inv_mod_part_2_d = inv_mod(part_2_d, MOD);
        ui part_2 = MVC((((part_2_a + part_2_b) % MOD) - part_2_c) % MOD, inv_mod_part_2_d, MOD);
        //cout << part_2 << endl;

        ui coeff = MVC(a, part_1, MOD) + MVC(L, part_2, MOD);
        coeff %= MOD;
        ans += MVC(MVC(coeff, digit, MOD), inv_mod_9, MOD);
        ans %= MOD;
    }
    return ans;
}

int main()
{
    auto primes = get_primes(15485864);
    auto C = concat_nums(primes);
    ui ans = S(C, K);
    if (ans < 0)
        ans += MOD;
    cout << ans << endl;
    return 0;
}
